Add support for a predefined list of party names in the config

osrscommits opened the pull request:
Add a config field for allowing the user to specify a list of party names. This is useful when a player needs to join on a bunch of accounts (alting), for players who join a different party based on who they're bossing with (eg., some friends are often in party x, others often in party y, etc.).

`ControlPanel#updateControls` has to to a bit extra shenanigans to handle updating the list of buttons; happy to move that to a helper function if preferred (or if any other changes). Also, there's some hardcoding about positioning with the `GridBagConstraints` (which was already there beforehand in the constructor, but is a bit more gross with the list logic now). Not sure if there's a better way, I'm not familiar with swing.

Screen recording from testing is too large to attach, so here's what I tested with this:

1. No buttons or header are shown when predefined party list config is empty (so for anyone who doesn't add names to the config option, there is no visual change to the sidebar)
2. Party buttons are updated when the config changes (new ones added and old ones deleted)
3. No extra buttons or header are shown when in a party
4. Trims whitespace before or after a comma-delimited entry. If there is a space within an entry, it's handled the same as a party with a space (gets converted to a hyphen). If it's a newline, it gets concatenated with no space (not really intentional but idk why there would be a newline in a party name)
    Very long party names just overflow and get cutoff in the sidebar; this is somewhere around 30 characters, so doesn't really seem like a problem to me (also not sure what the alternative would be).

osrscommits commented:
Whoops, didn't mean to open this just yet, will update description

osrscommits commented:
Realized I need to update this to handle whitespace in party names (I didn't know spaces were allowed)

ETA: Maybe they aren't? And the plugin just converts them to hyphens.

ETA v2: Handled, explained in the first comment

TheStonedTurtle commented:
Probably better to make this a dropdown instead of adding a bunch of new buttons?

osrscommits commented:
Thanks for taking a look! I find the list a bit jarring since choosing an option causes an action, rather than just being used as a selector, and personally prefer the buttons, but would be happy with either. Attached screenshots of both, if you prefer the dropdown there, I'll go ahead and update the PR.

<img width="218" height="267" alt="image" src="https://github.com/user-attachments/assets/3e2bf257-7347-4328-a421-720def7f13ce" />

<img width="216" height="254" alt="image" src="https://github.com/user-attachments/assets/6f0457c3-f371-4720-9170-ffdf6c52c54a" />

TheStonedTurtle commented:
I think it's fine having selecting the dropdown do something but maybe it is jarring to have it instantly join the party.

Maybe if this config exists the `Join previous party` button changes to `Join selected party` and the default option is `previous party`? Not sure how intuitive that would be.

Commit:
Add support for a predefined list of party names in the config

osrscommits commented:
Updated the code to use a dropdown. Some notes:

1. Only render the dropdown if there is data in the config field.
2. Added a custom renderer so the list can have the 'select a party' header, but not as a dropdown item. This also requires some extra handling in the listener and code updating the dropdown list (`isDropdownBeingUpdated`), as otherwise the dropdown will always select the first element after all elemnts are removed (and so was continuously joining the first party in the list). The way `isDropdownBeingUpdated` gets updated in `updateControls` feels a bit sus to me, as it's essentially a guard on the code block; maybe should be in a try-finally to ensure it always gets set back to false? Or, if there's a better way of doing the header in the list.

osrscommits commented:
The 'select an option from a list to join the party' did feel fine with some more testing; I prefer the buttons for not needing to scroll, but either is better than typing.

I think it's also good to keep the defaults simple (many ppl probably don't care/won't use about this feature), and so keeping it hidden/not modifying the other buttons if the config isn't provided works well for that.

osrscommits commented:
Anything else I need to change on this?

TheStonedTurtle commented on src/main/java/thestonedturtle/partypanel/ui/ControlsPanel.java:
```diff
@@ -25,18 +25,17 @@
-import java.awt.GridBagLayout;
-import java.awt.Insets;
-import java.awt.Toolkit;
+import java.awt.*;
```
Don't change these imports to `*` imports.
<img width="982" height="734" alt="image" src="https://github.com/user-attachments/assets/a1486545-2f19-4e88-94e1-c4549d0a1aee" />

TheStonedTurtle commented on src/main/java/thestonedturtle/partypanel/ui/ControlsPanel.java:
```diff
@@ -161,6 +190,29 @@ public void updateControls()
+			predefinedPartyNameDropdown.setVisible(!plugin.isInParty());
+			predefinedPartyNameDropdown.setSelectedIndex(-1);
+
+			c.gridx = 0;
```
Additionally if you add this to the constructor once you don't need to move `c = GridBag...` as a class level variable

TheStonedTurtle commented on src/main/java/thestonedturtle/partypanel/ui/ControlsPanel.java:
```diff
@@ -161,6 +190,29 @@ public void updateControls()
+		isDropdownBeingUpdated = true;
+		predefinedPartyNameDropdown.removeAllItems();
+		String predefinedPartyNames = plugin.getConfig().predefinedPartyNames();
+		if (predefinedPartyNames.isEmpty()) {
```
Did you try always adding this element in the constructor but with `setVisible(false)`? Then the only thing we need to do is updating the visibility depending on the state of the config option/party state?

Additionally, we may want to store the list from the config as a variable in here and only update the elements in this dropdown if something has actually changed. Could be as simple as:
```java
  private String currentlyDisplayedConfig =  "";

  ...
  if (currentlyDisplayedConfig != predefinedPartyNames) {
    // Rebuild the list
  }
```

This may also remove the need for `isDropdownBeingUpdated` unless adding new items causes them to be selected.

TheStonedTurtle reviewed (changes requested):
Sorry this review took me a while.
